src/NHibernate/Linq/Visitors/ParameterTypeLocator.cs

```diff
@@ -302,6 +302,7 @@ private bool TryLinkContainsMethod(QueryModel queryModel)
 				// ContainsResultOperator where the constant expression is dislocated from the related expression,
 				// we have to manually link the related expressions.
 				if (queryModel.ResultOperators.Count != 1 ||
+					queryModel.BodyClauses.Count > 0 ||
```

Comment:
We also have different issue regarding `Contains` on parameter constants. See #2872. In 5.3 we didn't handle it properly but exception is thrown.  
I wonder if this change makes `Where` part be silently ignored.

Reply:
I must admit I don't quite understand why TryLinkContainsMethod is needed.

Reply:
It's required for proper parameter type recognition(See #2459). So it's required for queries like `Where(order => collectionParameter.Contains(order.Id))` or `Where(order => order.Lines.Contains(parameter))`
So in theory your changes might break second type of queries (Contains on subquery) with Body clauses. But removing completely this check I see fails only for first type of queries (Contains on collection parameter). So it's either not properly covered in tests or this change not needed any more...

Reply:
> So in theory your changes might break second type of queries

So maybe to be on a safe side we should keep logic adding related expressions and instead replace final "return true" that skips further processing with something like:
```C#
//Subqueries not based on parameters need further processing
return right.NodeType != ExpressionType.Constant;
```

Reply:
My last suggestion is actually wrong. And it should be actually pretty safe as is... So marking it as resolved.